### DISCUSSION TITLE
Startup timing measurements unreliable in iOS15

### DISCUSSION THREAD

**Discussion by michalkalinowski-** [MobileNativeFoundation/discussions#146] [General]
Hello all,

I'm an iOS dev working on mobile apps at Yelp. We're tracking startup time of our apps and since the release of iOS15 we've started seeing large number of startup time events with unreasonably large durations. At this point the "large duration events" add up to almost 20% of all the events reported by iOS15 clients which is worrisome.

I've heard from devs working on other apps they see similar large duration events so I thought I'd reach out to the community to check if anyone has any clue about the root cause.

What we know so far:
* there's very large spread of outliers - they range from minutes to days (largest ones we've seen are ~12 days)
* _most likely_ they are not related first launch after the install (we 2+ "large duration events" for the same user within a day)
* we didn't manage to reproduce on a device over a course of a couple of days.

**Details of our perf measurement setup:**
* we start the clock as soon as our app hits the `main`
* we get timestamps from `mach_absolute_time()` which we assume is guaranteed to be a monotonic clock ([docs](https://developer.apple.com/documentation/kernel/1462446-mach_absolute_time)) 
* large timings (hours to days) are reported even on the earliest point we measure which is AppDelegate's `applicationDidFinishLaunchingWithOptions)

Is anyone else seeing similar noise in startup time measurements in their apps? Any clue on what's the root cause?

**Comment by mattrobmattrob** [comment]
If you're using a method similar to https://eisel.me/startup, it's likely caused by [`dyld` prewarming](https://twitter.com/_saagarjha/status/1431090656487763968?s=20). Michael recently added a section in the code snippet of that blog post about the implications of the iOS 15 pre-warming.

```ObjC
// With iOS 15 pre-warming, initializers and other pre-main steps are run preemptively, potentially hours before the
// app is started and main() is run. So, take the difference between process start time in pre-main initializer
// (which we control with this "constructor" attribute) and add that later with post-main time.
// NOTE: any initializers run after this will not have their startup time impact included. We try to make this
// initializer go first using a priority of "0", but there are no guarantees.
```

We are (at this point) just excluding measurements from iOS 15 until we have time to revisit calculating it in a more correct way.

**Comment by cwalo** [comment]
@michalkalinowski- Your findings match ours. We saw some extremely long tails in our app launch metrics for users on iOS 15. For now we've just been excluding durations longer than 10 minutes (kind of arbitrary). It's definitely a bummer that even with this workaround, initialization time isn't reliably captured. [Here's a good article on how this mechanism works](https://www.emergetools.com/blog/posts/iOS15LaunchTime) if you haven't seen it.

**Comment by ileitch** [comment]
@michalkalinowski- @cwalo I'm keen to clarify where exactly you're seeing this increase. As @mattrobmattrob mentioned, the `dyld` prewarming is a known optimization in iOS 15, but that _should_ only affect pre-main measurements. @michalkalinowski- you mentioned you only start timing in `main`, which suggests there may be another cause for the increase you're seeing.

**Reply by saagarjha** [reply]
Pre-warming technically executes in `UIApplicationMain`, not `main`, so if you're gathering metrics before that you'll be affected.

**Reply by cwalo** [reply]
Is that true? Maybe I'm misunderstanding. [According to docs:](https://developer.apple.com/documentation/uikit/app_and_environment/responding_to_the_launch_of_your_app/about_the_app_launch_sequence)

> Prewarming executes an app’s launch sequence up until, but not including, when main() calls UIApplicationMain.

**Reply by oxcug** [reply]
It would definitely be the other way around. Unless they re-wrote the way the kernel works recently, `UIApplicationMain` could be circumvented for prewarming, but not `main` since that's the defacto entry point executed by the kernel to boot up a new process.

**Reply by saagarjha** [reply]
Perhaps I phrased it confusingly: to be clear, what I mean is that iOS will call your `main`, including all the code you have in your `main` above `UIApplicationMain`. Then, you'll call `UIApplicationMain` and UIKit will check whether it needs to suspend inside of that.

**Comment by cwalo** [comment]
@ileitch We start measuring from process start time. @michalkalinowski- Ditto, can you clarify your launch span? Also curious to know how exactly you're using `mach_absolute_time`. I assume you're [converting it back to a real time interval](https://developer.apple.com/library/archive/qa/qa1398/_index.html)?

**Reply by michalkalinowski-** [reply]
We've set up our measurement to capture the time between `main` is called (we record `mach_absolute_time` in the first lines of main) till specific points in launch process like hitting `AppDidFinishLaunching`, hiding system splash screen etc.

You're correct assuming we convert the Int64 returned by `mach_absolute_time` similarly to how it is done in [those docs](https://developer.apple.com/library/archive/qa/qa1643/_index.html):
```
func durationBetween(_ first: UInt64, _ second: UInt64) -> Double {
    // gets timebase info
    // ...
    let nanoseconds = ((second - first) * UInt64(info.numer) / UInt64(info.denom))
    return Double(nanoseconds / NSEC_PER_MSEC)
  }
```

Even with the setup as outlined above we're seeing "large durations" reported as early as for "appDidFinishLaunching" event. The twitter thread you linked below is very helpful and sheds some light on what may be happening. This sadly does impact our current strategy of assessing startup performance quite a bit as it seems that our current metrics (main -> some signpost) will no longer be useful with pre-warmed apps.

**Comment by oxcug** [comment]
## Tl;dr
There's a few things that _could_ be happening here, but tl;dr, I would *highly* recommend AGAINST rolling your own solution. Use the perf data that [MetricKit.framework](https://developer.apple.com/documentation/metrickit/improving_your_app_s_performance/) (in production) and [XCTApplicationLaunch](https://developer.apple.com/documentation/xctest/xctapplicationlaunchmetric) (for UI Testing) provide.

They are the two most reliable data sources because they use the OS values from internal systems (e.g Springboard and Launchd) which govern the launch process. Those values are only exposed through this API (I know because I helped write them) and are the same one's used to do things like determine if the OS should kill your application.
 
## *For science only* ™️

The `uint64_t` value that is vended by`mach_absolute_time` could simply be overflowing (hence the huge delta from days to seconds) when you convert them in your process to calculate the duration. You can check this by using clang's [builtin](https://belaycpp.com/2021/06/08/dealing-with-integer-overflows/) functions for [detecting integer overflow](https://clang.llvm.org/docs/LanguageExtensions.html).

Another possibility is that the application is simply being killed on startup (e.g dyld_prewarm as previously mentioned) and the next time someone opens the application you're measurement is accurate in that the successful launch was X days / hours after. Off the top of my head, those would be my best guesses. But again, don't roll your own solution for measuring perf here, use MetricKit/XCTest.

**Reply by mattrobmattrob** [reply]
Has anyone published an example of using MetricKit to ingest data into internal monitoring/tracing/analytics tooling?

**Reply by sberrevoets** [reply]
At Lyft we're in the process of rolling this out. @cwalo might be able to give some bullet points on the approach.

**Reply by cwalo** [reply]
Our MetricKit implementation just landed and it is integrated into our internal analytics pipeline. I'll follow up as soon as I get a chance to run some queries against those events.

**Reply by michalkalinowski-** [reply]
I've just checked our MetricKit data and we only have 0.1% of startup events that are larger than 35s (which is our definition of "unrealistically large durations" for now). It looks like it's handling pre-warmed apps correctly then.

Note that MetricKit reports only the time till system splash screen is hidden so it's not very useful, at least for us. 
Looking forward to your findings @cwalo.

**Reply by oxcug** [reply]
@michalkalinowski- Quick correction, if memory serves, MetricKit measures from the moment you tap the app on the springboard (before main or your process is called) until applicationDidFinishLaunching has finished executing. The OS regards everything up till that point as part of the launch sequence.

**Reply by bpsrajat** [reply]
After seeing the same issues with our startup measurement, we are rolling out the metric kit as well.  
@cwalo  @sberrevoets @michalkalinowski Are you guys analyzing the [MXHistogram](https://developer.apple.com/documentation/metrickit/mxhistogram) that comes from the [MXAppLaunchMetric](https://developer.apple.com/documentation/metrickit/mxapplaunchmetric)?
is there a different way to get a single value (maybe avg.) launch time?

**Reply by mkj-is** [reply]
If you support tvOS then sadly using MetricKit and [XCTApplicationLaunch](https://developer.apple.com/documentation/xctest/xctapplicationlaunchmetric) is not an option.

- MetricKit is not available on tvOS ([FB13476055](http://www.openradar.me/FB13476055)).
- XCTApplicationLaunchMetric is not reported for tvOS ([FB11811540](http://www.openradar.me/FB11811540))

The good thing is that pre-warming is not present on tvOS, meaning using custom measurements is viable.

One more thing to note is that MetricKit is supported on visionOS. So, Apple TV is the only Apple platform (not taking watchOS into account) missing the support.

**Comment by cwalo** [comment]
This [twitter thread](https://twitter.com/FelixOmni/status/1466140731324739592?s=20) is also highly relevant

> We also burned a TSI on this issue, and it was confirmed to us that despite the documentation, pre-warming _does_ call AppDelegate functions including didFinishLaunchingWithOptions. We were further told that "the jury is still out" on whether it's a bug or intended behavior.
